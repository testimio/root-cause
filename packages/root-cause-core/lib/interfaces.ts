import type { Page as PuppeteerPage } from 'puppeteer';
import type { Page as PlaywrightPage } from 'playwright';
import type { TestContextInterface } from './TestContext';
import type { TestEndStatus } from './attachInterfaces';
import type { InstrumentedFunctionResult } from '@testim/root-cause-types';

export type RootCausePage = PuppeteerPage | PlaywrightPage;

export interface AttachOptions {
  noResultsServer?: boolean;
  externalTestName?: string;
  resultsDirectory?: string;
}

export interface CommonEachHookArgs {
  testContext: TestContextInterface;
  fnName: string;
  proxyContext: any;
  rootPage: RootCausePage;
  args: any[];
  methodCallData: ProxiedMethodCallData[];
}

export type BeforeHookArgs = CommonEachHookArgs;

export type BeforeHook = (args: BeforeHookArgs) => Promise<void>;

export interface AfterHookArgs extends CommonEachHookArgs {
  instrumentedFunctionResult: InstrumentedFunctionResult<any, any>;
}

export type AfterHook = (args: AfterHookArgs) => Promise<void>;

export interface BeforeAllHookArgs {
  testContext: TestContextInterface;
  proxyContext: any;
  rootPage: RootCausePage;
}

export type BeforeAllHook = (args: BeforeAllHookArgs) => Promise<void>;

export interface AfterAllHookArgs {
  testContext: TestContextInterface;
  endStatus: TestEndStatus<unknown, unknown>;
}

export type AfterAllHook = (args: AfterAllHookArgs) => Promise<void>;

export interface IAutomationFrameworkInstrumentor {
  // register additional hooks on a context, this is implemented for
  // puppeteer/selenium
  // registerHooks(page: T, hooks: AutomationFrameworkHooks<T>): T;
  registerBeforeAllHook(hook: BeforeAllHook): void;
  registerAfterAllHook(hook: AfterAllHook): void;
  registerBeforeHook(hook: BeforeHook): void;
  registerAfterHook(hook: AfterHook): void;
  wrapWithProxy<T extends object>(proxiedObject: T): T;
  end(endStatus: TestEndStatus<unknown, unknown>): Promise<void>;
  pause(): void;
  resume(): void;
}

export interface ProxiedMethodCallData {
  creationFunction: string;
  selector?: string;
  text?: string;
  index?: number;
}
