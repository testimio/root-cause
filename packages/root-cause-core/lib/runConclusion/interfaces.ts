import { TestResultFile } from '@testim/root-cause-types';

export interface RootCauseRunResultEntry {
  singleResultDir: string;
  id: string;
  testData: TestResultFile;
}
