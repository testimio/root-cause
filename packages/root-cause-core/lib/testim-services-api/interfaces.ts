// generated with json2ts

interface TestimExecutionExecutionRunConfig {
  parallel: number;
  browser: string;
  runnerVersion: string;
  testimBranch: string;
  canaryMode: boolean;
  source: string;
  testPlans: any[];
  testLabels: any[];
  testNames: any[];
  testIds: any[];
  testConfigs: any[];
  testConfigIds: any[];
  browserTimeout: number;
  timeout: number;
  newBrowserWaitTimeout: number;
  tunnelPort: string;
  runnerMode: string;
  sessionType: string;
  companyId: string;
  testData?: any;
  testDataTotal?: any;
  testDataIndex?: any;
  testConfig: {}; // not too interesting
}

interface TestExecution {
  testId: string;
  status: string;
  name: string;
  resultId: string;
  isTestsContainer: boolean;
  testRunOverrideStatus: string;
  config: TestimExecutionExecutionRunConfig;
  success?: boolean;
  reason?: string;
  childTestResultIds?: string[];
  startTime: number;
  endTime: number;
  parentResultId?: string;
  show?: boolean;
}

export interface TestimBackendExecutionFormatSubsetForReporting {
  runId: string;
  execution: {
    [id: string]: {
      testId: string;
      name: string;
      success?: boolean;
      resultId: string;
      parentResultId?: string;
      isTestsContainer: boolean;
    };
  };
  startTime: number;
}
export interface TestimBackendExecutionInputFormat {
  runId: string;
  projectId: string;
  labels: string;
  startTime: number;
  endTime: number;
  execution: { [id: string]: TestExecution };
  status: string;
  config: TestimExecutionExecutionRunConfig;
  resultLabels: string[];
  remoteRunId: string;
  metadata: any;
  show: boolean;
}
