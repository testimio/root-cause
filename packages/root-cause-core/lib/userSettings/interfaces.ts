export interface PossibleUserSettings {
  features?: Partial<FeaturesSettings>;
}

export interface FeaturesSettings {
  /**
   * Take screenshot for every step
   *
   * Default to true, with jpeg 85
   */
  screenshots?: boolean | ScreenshotsSettings;

  /**
   * Record the browser's console
   *
   * @default true
   */
  console?: boolean;

  /**
   * Record network activity, make it available in the UI and as har file
   */
  networkLogs?: boolean;

  /**
   * When using jest integration, record jest expect calls as assertion steps
   */
  jestAssertions?: boolean;

  /**
   * Record the DOM content of the page for every step
   */
  html?: boolean;
}

export interface ResolvedSettings {
  features: {
    screenshots: false | Required<ScreenshotsSettings>;
    console: boolean;
    networkLogs: boolean;
    jestAssertions: boolean;
    html: boolean;
  };
}

export interface ScreenshotsSettings {
  format: 'png' | 'jpeg';

  /**
   * Quality of jpeg image, from 1 to 100
   *
   * @default 85
   */
  quality?: number;

  /**
   * When true, takes a screenshot of the full scrollable page, instead of the currently visible
   *
   * @default false
   */
  fullPage?: boolean;
}

type ContentType = string;

/**
 * placeholder, not in use yet
 */
export interface NetworkLogsSettings {
  /**
   * Save the response bodies into the har file, or only for specific content types
   *
   * @default false
   */
  includeResponseBody?: boolean | ContentType[];
}
