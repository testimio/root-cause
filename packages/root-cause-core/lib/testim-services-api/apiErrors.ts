type ApiErrorCode =
  | 'E_UNAUTHENTICATED'
  | 'E_CREDENTIALS_DONT_MATCH'
  | 'E_CREATING_EXECUTION_FAILED'
  | 'E_UPLOAD_REQUEST_FAILED'
  | 'E_FILE_NOT_FOUND';
class TestimApiErrorBase extends Error {
  constructor(public code: ApiErrorCode, message: string) {
    super(message);
  }
}
export class TestimApiError extends TestimApiErrorBase {
  constructor(private originalError: Error, message: string, code: ApiErrorCode) {
    super(code, message);
  }
}
export class TestimNotFoundError extends TestimApiErrorBase {
  constructor(message: string) {
    super('E_FILE_NOT_FOUND', message);
  }
}
