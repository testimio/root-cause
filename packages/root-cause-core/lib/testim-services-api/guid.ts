'use strict';

const a = 'qwertyuiopasdfghjklzxcvbnmQWERTYUIOPASDFGHJKLZXCVBNM1234567890';

export function guid(n = 10) {
  let str = '';
  for (let i = 0; i < n; i++) {
    const index = Math.floor(Math.random() * (a.length - 1));
    str += a[index];
  }
  return str;
}
