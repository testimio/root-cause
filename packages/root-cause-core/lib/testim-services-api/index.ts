import type nodeFetch from 'node-fetch';

import { TestimAuthApi } from './testimAuthApi';
import { TestimExecutionsApi } from './testimExecutionsApi';
import { TestimAssetsApi } from './testimAssetsApi';
import { RootCauseExecutionsApi } from './rootCauseExecutionsApi';
import type { AbortSignal } from 'abort-controller';

export { TestimAuthApi } from './testimAuthApi';
export { TestimExecutionsApi } from './testimExecutionsApi';
export { TestimAssetsApi } from './testimAssetsApi';
export { RootCauseExecutionsApi } from './rootCauseExecutionsApi';

/**
 * This class is a facade over the Testim Backend API
 */
export class TestimApi {
    auth: TestimAuthApi;
    assets: TestimAssetsApi;
    executions: TestimExecutionsApi;
    executionsApi: RootCauseExecutionsApi;
    projectId?: string;
    ciToken?: string;
    constructor(baseUrl?: string, fetch?: typeof nodeFetch) {
        this.auth = new TestimAuthApi(baseUrl, fetch);
        this.assets = new TestimAssetsApi(baseUrl, this.auth, undefined, fetch);
        this.executions = new TestimExecutionsApi(baseUrl, this.auth, fetch);
        this.executionsApi = new RootCauseExecutionsApi(this.executions, this.assets);
    }
    async authenticate(projectId: string, ciToken: string, signal?: AbortSignal) {
        await this.auth.getToken(projectId, ciToken, signal);
        this.projectId = projectId;
        this.ciToken = ciToken;
    }
}
