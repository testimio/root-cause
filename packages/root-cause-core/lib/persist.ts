import { TestimApi } from './testim-services-api';
import { getTestimCredentials } from './testim-credentials-manager';
import { resolve as pathResolve } from 'path';
import { constructResultDir } from './utils';
import fs from 'fs-extra';
import ora from 'ora';
import untildify from 'untildify';
import { TestimBackendExecutionInputFormat, TestimBackendExecutionFormatSubsetForReporting } from './testim-services-api/interfaces';

export async function persist(runId: string, {
    projectRoot, resultLabel, project, token, jUnitReportPath,
}: { projectRoot?: string; resultLabel: string[]; project?: string; token?: string; jUnitReportPath?: string }) {
    if (!runId) {
        console.log('Please pass a valid runId');
        return;
    }
    const credentials = (project && token) ? { projectId: project, ciToken: token } : await getTestimCredentials();
    if (!jUnitReportPath && process.env.TESTIM_PROJECT_JUNIT_PATH) {
        jUnitReportPath = untildify(process.env.TESTIM_PROJECT_JUNIT_PATH);
    }
    if (!credentials.ciToken || !credentials.projectId) {
        console.log('No Root Cause Credentials found, please make sure the TESTIM_PROJECT_ID' +
        ' and TESTIM_PROJECT_TOKEN environment variables are defined.');
        return;
    }
    const api = new TestimApi(process.env.SERVICES_URL);
    await api.authenticate(credentials.projectId, credentials.ciToken);
    const dir = constructResultDir(projectRoot || process.cwd());
    const path = pathResolve(dir, 'runs', runId);

    if (!await fs.pathExists(path)) {
        console.log('Run does not exist, looked at', path);
        return;
    }
    const spinner = ora(`UpÂ´ tests for suite run ${runId}`).start();
    try {
        if (!Array.isArray(resultLabel)) {
            resultLabel = [];
        }
        resultLabel.push('RootCause');
        const { testimFormat } = await api.executionsApi.createExecution(path, {
            projectId: credentials.projectId,
            resultLabels: resultLabel,
            // TODO(Benji) discuss with Elad, companyId is only used for
            // the testConfig and is probably not important?
            companyId: '',
        }, (progress: number) => {
            const text = `${(progress * 100).toFixed(2)}%`;
            spinner.text = `Uploading tests for suite run ${runId} (${text})`;
        });
        spinner.succeed('Execution Created Successfully');
        try {
            if (jUnitReportPath) {
                await saveJunitReport({
                    jUnitReportPath, testimExecution: testimFormat, projectId: credentials.projectId,
                });
                console.log('Report file saved at ', jUnitReportPath);
            }
        } catch (e) {
            console.error('Error generating jUnit Report', e);
        }
        spinner.succeed(`Watch your tests in https://app.testim.io/#/project/${credentials.projectId}/runs/suites/${testimFormat.runId}`);
    } catch (e) {
        spinner.fail(`Error Creating Execution: ${e && e.message}`);
        console.error(e);
    }
}

async function saveJunitReport({
    jUnitReportPath, testimExecution, projectId,
}: { jUnitReportPath: string; testimExecution: TestimBackendExecutionInputFormat; projectId: string }) {
    const report = generateJunitReport({ testimExecution, projectId });
    await fs.writeFile(jUnitReportPath, report);
}

export function generateJunitReport({
    testimExecution, projectId,
}: { testimExecution: TestimBackendExecutionFormatSubsetForReporting; projectId: string }) {
    const actualTests = Object.values(testimExecution.execution).filter(test => !test.isTestsContainer);
    const dateValue = new Date(testimExecution.startTime).toUTCString();
    const files = Object.values(testimExecution.execution).filter(test => test.isTestsContainer);
    const baseUrl = `https://app.testim.io/#/project/${projectId}/branch/master/root-cause/run/${testimExecution.runId}`;
    const attr = (str: string) => `"${str.replace(/"/g, '&quot').replace(/&/g, '&amp').replace(/</, '&lt;')}"`;
    // Includes the right values (from) https://llg.cubic.org/docs/junit/ with the battle tested ones (our cli jUnitReporter)
    const report = `<?xml version="1.0" encoding="UTF-8" standalone="yes"?><testsuites disabled='0'>${files.map(file => {
        const tests = actualTests.filter(x => x.parentResultId === file.resultId);
        const failed = tests.filter(x => !x.success);
        return `\t<testsuite name=${attr(file.name)} tests="${tests.length}" failure="${failed.length}" failures="${failed.length}" timestamp="${dateValue}">
                 ${tests.map(test => `<testcase name=${attr(test.name)} classname="rootcause.test"><system-out>${baseUrl}/test/${test.testId}</system-out></testcase>`).join('\n\t\t\t\t')}
             </testsuite>`;
    }).join('\n\t\t\t')}
         <system-out>Suite Run URL: ${baseUrl}</system-out>
    </testsuites>`;

    // const report = `<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
    // <testsuites>
    //   <testsuite name="playground TES-6699-demo:playground with config '1280x800'" tests="8" failure="0" timestamp="2020-07-17T12:20:08.810Z">
    //     <testcase name="playground - validate generated playwright code on record" classname="testim.io.test" time="27.812">
    //       <system-out>https://app.testim.io/#/project/Testim-UI_f93f4936-f415-460a-b9a8-0c2ac81e7309/branch/master/test/LN05PyoNg8?result-id=kbePYrZuNZ</system-out>
    //     </testcase>
    //     <testcase name="playground - validate record and play buttons text change and functionality" classname="testim.io.test" time="22.414">
    //       <system-out>https://app.testim.io/#/project/Testim-UI_f93f4936-f415-460a-b9a8-0c2ac81e7309/branch/master/test/g5tZB7pIal?result-id=LunODjZShG</system-out>
    //     </testcase>
    //     <testcase name="playground - validate successful playback in the playground editor" classname="testim.io.test" time="36.369">
    //       <system-out>https://app.testim.io/#/project/Testim-UI_f93f4936-f415-460a-b9a8-0c2ac81e7309/branch/master/test/exGG3hJZ3U?result-id=SmpWvHUaVb</system-out>
    //     </testcase>
    //     <testcase name="playground - validate playwright code updates when updating codeless editor" classname="testim.io.test" time="38.016">
    //       <system-out>https://app.testim.io/#/project/Testim-UI_f93f4936-f415-460a-b9a8-0c2ac81e7309/branch/master/test/oharvsUAd3?result-id=cr8UXd1IUk</system-out>
    //     </testcase>
    //     <testcase name="playground - validate that red recording banner is shown when recording and hidden when stopping record" classname="testim.io.test" time="19.697">
    //       <system-out>https://app.testim.io/#/project/Testim-UI_f93f4936-f415-460a-b9a8-0c2ac81e7309/branch/master/test/EHLWRqG7Cv?result-id=DsR8blgaHt</system-out>
    //     </testcase>
    //     <testcase name="playground - validate that steps are being recorded in the codeless editor" classname="testim.io.test" time="24.006">
    //       <system-out>https://app.testim.io/#/project/Testim-UI_f93f4936-f415-460a-b9a8-0c2ac81e7309/branch/master/test/fNCHJoNETT?result-id=7f4yC3okpX</system-out>
    //     </testcase>
    //     <testcase name="playground - validated that editing a target element in the codeless editor updates the selector in playwright" classname="testim.io.test" time="31.326">
    //       <system-out>https://app.testim.io/#/project/Testim-UI_f93f4936-f415-460a-b9a8-0c2ac81e7309/branch/master/test/Jiim2YdpFQ?result-id=acza9OnipN</system-out>
    //     </testcase>
    //     <testcase name="playground - validate images are being uploaded successfully from blob" classname="testim.io.test" time="49.895">
    //       <system-out>https://app.testim.io/#/project/Testim-UI_f93f4936-f415-460a-b9a8-0c2ac81e7309/branch/master/test/Kgx42VhQ2Q?result-id=vSdlNR4yg1</system-out>
    //     </testcase>
    //   </testsuite>
    // </testsuites>`;
    return report;
}

