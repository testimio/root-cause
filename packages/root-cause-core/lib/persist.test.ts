import { generateJunitReport } from './persist';
import { TestimBackendExecutionFormatSubsetForReporting } from './testim-services-api/interfaces';

const dummyProjectId = 'brave-dog';

function makeExampleExecution(
  name = 'Example Execution',
  execution: TestimBackendExecutionFormatSubsetForReporting['execution'] = {}
): TestimBackendExecutionFormatSubsetForReporting {
  return {
    runId: '1',
    execution,
    startTime: 1,
  };
}
describe('the jUnit reporter for cloud executions', () => {
  it('generates an empty report if there are no tests', () => {
    const report = generateJunitReport({
      projectId: dummyProjectId,
      testimExecution: makeExampleExecution(),
    });
    expect(report).toMatchSnapshot();
  });
  it('generates a report with one test if the execution contains a single test', () => {
    const report = generateJunitReport({
      projectId: dummyProjectId,
      testimExecution: makeExampleExecution('Example Execution One Test', {
        resultId: {
          testId: 'testId',
          name: 'some test',
          success: true,
          resultId: 'someResultId',
          parentResultId: 'parentResultId',
          isTestsContainer: false,
        },
        parentResultId: {
          testId: 'parentTestId',
          name: 'some test',
          success: true,
          resultId: 'someResultId',
          isTestsContainer: true,
        },
      }),
    });
    expect(report).toMatchSnapshot();
  });
  it('generates a report with several tests if the execution contains several tests', () => {
    const report = generateJunitReport({
      projectId: dummyProjectId,
      testimExecution: makeExampleExecution('Example Execution Two Tests', {
        resultId: {
          testId: 'testId',
          name: 'some test',
          success: true,
          resultId: 'someResultId',
          parentResultId: 'parentResultId',
          isTestsContainer: false,
        },
        resultId2: {
          testId: 'testId2',
          name: 'some other test',
          success: true,
          resultId: 'someResultId',
          parentResultId: 'parentResultId',
          isTestsContainer: false,
        },
        parentResultId: {
          testId: 'parentTestId',
          name: 'some test',
          success: true,
          resultId: 'someResultId',
          isTestsContainer: true,
        },
      }),
    });
    expect(report).toMatchSnapshot();
  });
  it('generates a report with failed tests if a test failed', () => {
    const report = generateJunitReport({
      projectId: dummyProjectId,
      testimExecution: makeExampleExecution('Example Execution One Test', {
        resultId: {
          testId: 'testId',
          name: 'some test',
          success: false,
          resultId: 'someResultId',
          parentResultId: 'parentResultId',
          isTestsContainer: false,
        },
        parentResultId: {
          testId: 'parentTestId',
          name: 'some test',
          success: false,
          resultId: 'someResultId',
          isTestsContainer: true,
        },
      }),
    });
    expect(report).toMatchSnapshot();
  });
});
