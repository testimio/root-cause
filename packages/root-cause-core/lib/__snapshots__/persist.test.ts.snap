// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the jUnit reporter for cloud executions generates a report with failed tests if a test failed 1`] = `
"<?xml version=\\"1.0\\" encoding=\\"UTF-8\\" standalone=\\"yes\\"?><testsuites disabled='0'>	<testsuite name=\\"some test\\" tests=\\"0\\" failure=\\"0\\" failures=\\"0\\" timestamp=\\"Thu, 01 Jan 1970 00:00:00 GMT\\">
                 
             </testsuite>
         <system-out>Suite Run URL: https://app.testim.io/#/project/brave-dog/branch/master/root-cause/run/1</system-out>
    </testsuites>"
`;

exports[`the jUnit reporter for cloud executions generates a report with one test if the execution contains a single test 1`] = `
"<?xml version=\\"1.0\\" encoding=\\"UTF-8\\" standalone=\\"yes\\"?><testsuites disabled='0'>	<testsuite name=\\"some test\\" tests=\\"0\\" failure=\\"0\\" failures=\\"0\\" timestamp=\\"Thu, 01 Jan 1970 00:00:00 GMT\\">
                 
             </testsuite>
         <system-out>Suite Run URL: https://app.testim.io/#/project/brave-dog/branch/master/root-cause/run/1</system-out>
    </testsuites>"
`;

exports[`the jUnit reporter for cloud executions generates a report with several tests if the execution contains several tests 1`] = `
"<?xml version=\\"1.0\\" encoding=\\"UTF-8\\" standalone=\\"yes\\"?><testsuites disabled='0'>	<testsuite name=\\"some test\\" tests=\\"0\\" failure=\\"0\\" failures=\\"0\\" timestamp=\\"Thu, 01 Jan 1970 00:00:00 GMT\\">
                 
             </testsuite>
         <system-out>Suite Run URL: https://app.testim.io/#/project/brave-dog/branch/master/root-cause/run/1</system-out>
    </testsuites>"
`;

exports[`the jUnit reporter for cloud executions generates an empty report if there are no tests 1`] = `
"<?xml version=\\"1.0\\" encoding=\\"UTF-8\\" standalone=\\"yes\\"?><testsuites disabled='0'>
         <system-out>Suite Run URL: https://app.testim.io/#/project/brave-dog/branch/master/root-cause/run/1</system-out>
    </testsuites>"
`;
